src/rank.c

```diff
@@ -0,0 +1,253 @@
+#include <rlang.h>
+#include "vctrs.h"
+#include "equal.h"
+#include "order-radix.h"
+#include "decl/rank-decl.h"
+
+// [[ register() ]]
+r_obj* vctrs_rank(r_obj* x,
+                  r_obj* ties,
+                  r_obj* na_propagate,
+                  r_obj* direction,
+                  r_obj* na_value,
+                  r_obj* nan_distinct,
+                  r_obj* chr_transform) {
+  const enum ties c_ties = parse_ties(ties);
+  const bool c_na_propagate = r_bool_as_int(na_propagate);
+  const bool c_nan_distinct = r_bool_as_int(nan_distinct);
+
+  return vec_rank(
+    x,
+    c_ties,
+    c_na_propagate,
+    direction,
+    na_value,
+    c_nan_distinct,
+    chr_transform
+  );
+}
+
+static
+r_obj* vec_rank(r_obj* x,
+                enum ties ties_type,
+                bool na_propagate,
+                r_obj* direction,
+                r_obj* na_value,
+                bool nan_distinct,
+                r_obj* chr_transform) {
+  r_ssize size = vec_size(x);
+
+  r_keep_t pi_x;
+  KEEP_HERE(x, &pi_x);
+
+  r_obj* missing = r_null;
+  r_keep_t pi_missing;
+  KEEP_HERE(missing, &pi_missing);
+  int* v_missing = NULL;
+
+  r_obj* not_missing = r_null;
+  r_keep_t pi_not_missing;
+  KEEP_HERE(not_missing, &pi_not_missing);
+  int* v_not_missing = NULL;
+
+  r_ssize rank_size = -1;
+
+  if (na_propagate) {
+    // Slice out non-missing values of `x` to rank.
+    // Retain `non_missing` logical vector for constructing `out`.
+    missing = vec_equal_na(x);
+    KEEP_AT(missing, pi_missing);
+    v_missing = r_lgl_begin(missing);
+
+    bool any_missing = r_lgl_any(missing);
+    if (!any_missing) {
+      na_propagate = false;
+      goto skip_propagate;
+    }
+
+    for (r_ssize i = 0; i < size; ++i) {
+      v_missing[i] = !v_missing[i];
+    }
+
+    not_missing = missing;
+    KEEP_AT(not_missing, pi_not_missing);
+    v_not_missing = v_missing;
+    missing = NULL;
+    v_missing = NULL;
+
+    x = vec_slice(x, not_missing);
+    KEEP_AT(x, pi_x);
+
+    rank_size = vec_size(x);
+  } else {
+    skip_propagate:;
```

Comment:
hmm in this case I'm not sure I like the goto versus a slightly more verbose branching but feel free to leave as is.

Style: goto target should be half-indented and doesn't need a semi-colon (that actually makes it harder to spot that it's a goto target).

Reply:
I'm going to remove the goto